core/src/test/scala/unit/kafka/integration/PartitionLoggingTest.scala

```diff
@@ -56,6 +59,43 @@ class PartitionLoggingTest extends IntegrationTestHarness{
     waitForPartitionUnderMinISRState(tp, false)
   }
 
+  @Test
+  def loggingOfflinePartitionTest(): Unit = {
+    val controllerId = TestUtils.waitUntilControllerElected(zkClient)
+    // create the test topic with three replicas
+    val topicConfig = new Properties()
+    topicConfig.put(TopicConfig.MIN_IN_SYNC_REPLICAS_CONFIG, "2")
+    var nonControllerBrokerIds = servers.filter(s => s.config.brokerId != controllerId).map(_.config.brokerId)
+    if (nonControllerBrokerIds.size > 3) {
+      nonControllerBrokerIds = nonControllerBrokerIds.take(3)
+    }
+    assertEquals(3, nonControllerBrokerIds.size)
+    val topicName = "test_3replicas"
+    createTopic(topicName, Map(0 -> nonControllerBrokerIds), topicConfig)
+    val tp = new TopicPartition(topicName, 0)
+    checkPartitionOfflineStatus(false, tp)
+
+    // get the brokers that have the replicas
+    val partitionStateOpt = zkClient.getTopicPartitionState(tp)
+    assertTrue(partitionStateOpt.isDefined, "partitionState should exist for partition [test_3replicas,0]")
+    assertEquals(3, partitionStateOpt.get.leaderAndIsr.isr.size, "isr should be 3 for partition [test_3replicas,0]")
+    val brokersWithReplica = partitionStateOpt.get.leaderAndIsr.isr
+
+    // shutdown brokers to put the partition offline
+    val brokersToShutdown = servers.filter{s => brokersWithReplica.contains(s.config.brokerId)}
```

Comment:
I didn't spot any explicit changes to the default integration-test setup, which means you have three brokers in the cluster (one of which also acts as the controller) and are potentially shutting down the entire cluster here, right?

While that's also an interesting case, I think you should make the main test case a little closer to reality, i.e., the controller remains running while some or all of the data brokers go down.

Reply:
There are 4 brokers in this test "override protected def brokerCount: Int = 4". And I created the topic on brokers that are not controller (createTopic(topicName, Map(0 -> nonControllerBrokerIds), topicConfig)), so only non-controller brokers are shutdown. I am very sure on that the controller is not shutdown in the test because if the controller is down the offline partition count metric would not be updated.

Reply:
OK, very good. Thanks!